contracts/contracts/Tally.sol

```diff
@@ -154,59 +116,60 @@ contract Tally is Ownable, SnarkCommon, CommonUtilities, Hasher, DomainObjs {
     uint256 tallyBatchSize = TREE_ARITY ** intStateTreeDepth;
     uint256 batchStartIndex = tallyBatchNum * tallyBatchSize;
 
-    // save some gas because we won't overflow uint256
-    unchecked {
-      tallyBatchNum++;
-    }
-
     (uint256 numSignUps, ) = poll.numSignUpsAndMessages();
 
     // Require that there are untallied ballots left
     if (batchStartIndex >= numSignUps) {
       revert AllBallotsTallied();
     }
 
-    bool isValid = verifyTallyProof(_proof, numSignUps, batchStartIndex, tallyBatchSize, _newTallyCommitment);
+    uint256[] memory circuitPublicInputs = getPublicCircuitInputs(_newTallyCommitment);
+
+    // save some gas because we won't overflow uint256
+    unchecked {
+      tallyBatchNum++;
+    }
 
-    if (!isValid) {
+    if (!verifyTallyProof(circuitPublicInputs, _proof)) {
       revert InvalidTallyVotesProof();
     }
 
     // Update the tally commitment and the tally batch num
     tallyCommitment = _newTallyCommitment;
   }
 
+  /// @notice Get public circuit inputs.
+  /// @param _newTallyCommitment the new tally commitment to be verified
+  /// @return publicInputs public circuit inputs
+  function getPublicCircuitInputs(uint256 _newTallyCommitment) public view returns (uint256[] memory publicInputs) {
+    (uint256 numSignUps, ) = poll.numSignUpsAndMessages();
+    (uint8 intStateTreeDepth, , , ) = poll.treeDepths();
+    uint256 tallyBatchSize = TREE_ARITY ** intStateTreeDepth;
+    uint256 batchStartIndex = tallyBatchNum * tallyBatchSize;
+
+    publicInputs = new uint256[](5);
+    publicInputs[0] = sbCommitment;
+    publicInputs[1] = tallyCommitment;
+    publicInputs[2] = _newTallyCommitment;
+    publicInputs[3] = batchStartIndex;
+    publicInputs[4] = numSignUps;
+  }
+
   /// @notice Verify the tally proof using the verifying key
+  /// @param _publicInputs circuit public inputs
   /// @param _proof the proof generated after processing all messages
-  /// @param _numSignUps number of signups for a given poll
-  /// @param _batchStartIndex the number of batches multiplied by the size of the batch
-  /// @param _tallyBatchSize batch size for the tally
-  /// @param _newTallyCommitment the tally commitment to be verified at a given batch index
   /// @return isValid whether the proof is valid
   function verifyTallyProof(
-    uint256[8] calldata _proof,
-    uint256 _numSignUps,
-    uint256 _batchStartIndex,
-    uint256 _tallyBatchSize,
-    uint256 _newTallyCommitment
+    uint256[] memory _publicInputs,
```

Comment:
given we know how many inputs, maybe we constraint the array to a fixed size?

Reply:
I can try to check it, but my idea is: if we want to add more params, we don't need to change it later.

Reply:
So, it won't work because verifier needs dynamic size.